blog : Using Node-RED as an ETL tool (#2175)
FlowFuse/website: src/blog/2024/06/using-node-red-as-an-etl-tool.md

---

> title: Using Node-RED as an ETL Tool.

Reviewer: ```suggestion
title: Using Node-RED as an ETL Tool
```

> subtitle: A Step-by-Step Guide on How to Use Node-RED as an ETL Tool.

Reviewer: ```suggestion
subtitle: A Step-by-Step Guide to use Use Node-RED for your industrial ETL Tool.
```

> description: Learn how to utilize Node-RED as an ETL (Extract, Transform, Load) tool, and explore its advantages and disadvantages in this context.

Reviewer: ```suggestion
description: Learn how to utilize Node-RED as an ETL (Extract, Transform, Load) tool for your data industrial data ops.
```

date: 2024-06-05
authors: ["sumit-shinde"]
image:
tags:
  - etl
  - extract
  - transform
  - load
  - analysis
  - business intelligence
---

> ETL (Extract, Transform, Load) is essential for integrating and analyzing data, helping businesses unlock detailed insights. You already know Node-RED for its user-friendly approach to creating IoT applications, but did you know it can also be a powerful tool for ETL tasks? When IBM published a blog about using Node-RED for ETL, it caught a lot of attention and got people talking about its potential in this space. In this guide, we'll walk you through how to use Node-RED for ETL, sharing its strengths and weaknesses along the way.

Reviewer: @sumitshinde-84 This feels very ChatGPT, can you tone down the bravado?

<!--more-->

## What is ETL

> [ETL](https://en.wikipedia.org/wiki/Extract,_transform,_load) (Extract, Transform, Load) is a fundamental process in managing data efficiently. It consists of three main stages: extraction, transformation, and loading. In the extraction phase, data is gathered from various sources such as databases, APIs, or files. Then, in the transformation stage, the data is converted into a format that is suitable for analysis by cleaning, filtering, and structuring it. Finally, in the loading stage, the transformed data is stored in a target database or data warehouse for further analysis and decision-making. ETL plays a significant role in ensuring that data is organized, consistent, and readily available for generating insights and making informed decisions in businesses and organizations.

Reviewer: > from various sources such as databases, APIs, or files

@sumitshinde-84 Can you add more OT technology here?

## Node-RED as an ETL tool

> Node-RED is specifically used for building various types of IoT applications. During the development of such applications, we sometimes used Node-RED as an ETL is tool without even realizing it. For instance, when monitoring sensors across a facility, the process involves extracting data from sensors, transforming (cleaning and formatting) sensor data, and loading it into a database. There are thousands of other such examples. This demonstrates that Node-RED has already been adopted as an ETL tool at a certain level.

Reviewer: ```suggestion
Node-RED is specifically used for building various types of IoT applications. Data ops workload are part of these use-cases, and are often deployed without thinking of them as DataOps. For instance, when monitoring sensors across a facility, the process involves extracting data from sensors, transforming (cleaning and formatting) sensor data, and loading it into a database. There are thousands of other such examples.
```

*Note: We have mentioned only a few nodes that can be used in the ETL process, but there are thousands of other nodes that you can explore yourself by visiting the [Node-RED Nodes Library](https://flows.nodered.org/search?type=node)*

### Extracting

> Node-RED can extract data from various sources, including APIs, databases, local filesystems, and IoT devices using built-in nodes and community-contributed nodes. For example, the HTTP request node can be used to pull data from web services, while nodes for MySQL, MongoDB, and PostgreSQL can extract data from databases. Nodes for MQTT and Kafka can fetch data from message brokers. File nodes enable extraction of data from local filesystems, while different cloud nodes for platforms like AWS, GCP, and IBM Watson allow extraction of data from cloud storage services. Moreover, Node-RED running on edge devices can extract data from sensors connected directly to them.

Reviewer: Modbus, OPC-UA, etc etc aren't mentioned. Please lean into the industrial applicability more

### Transforming

> Transforming data involves cleaning, structuring, and applying business rules. This can be done using function nodes with custom JavaScript, along with nodes like JSON, Split, Join, Change, CSV, and XML for parsing and formatting.

Reviewer: I think this needs more details. The transformation is a key part, and the bread and butter of Node-RED.

### Visualizing and Analyzing

In addition to the core ETL process, visualizing and analyzing data is a crucial step for gaining insights. While visualization isn't part of the ETL acronym itself, it plays a significant role in understanding and interpreting transformed data. In Node-RED, you can easily visualize data using Node-RED Dashboard 2.0 charts.

Node-RED Dashboard 2.0 provides several chart options, and more are expected in the future. For more information and to explore available chart types, refer to the [Node-RED Dashboard 2.0 charts documentation](https://dashboard.flowfuse.com/nodes/widgets/ui-chart.html).

## Benefits of using Node-RED as an ETL tool.

1. **Ease of Use:** Its visual programming interface makes it accessible to non-developers.
2. **Flexibility:** A wide range of nodes and the ability to write custom JavaScript allow for flexible data processing.
3. **Integration:** Node-RED excels in integrating IoT devices and handling real-time data, making it well-suited for combining diverse data sources into unified workflows.
4. **Cost-Effective:** Being open-source, Node-RED can be a cost-effective alternative to expensive ETL tools.

> 5. **Community Support:** A large community provides a wealth of nodes, examples, and support.

Reviewer: ChatGPT? Feels uninspired and machine generated.

## Transforming Data

### Data Cleaning

1. Drag a Link-in node onto the canvas, double-click on it, and select the name of the Link-out node added in the above section.
2. Drag a Split node onto the canvas.
3. Drag a Change node onto the canvas and Set `msg.payload.full_name` to `payload.first_name & " " & payload.last_name` as the JSONata expression, this JSONata expression will concat the firstname and lastname properties, then delete unnecessary properties like `first_name` and `last_name`.
4. Drag a Join node onto the canvas.
5. Connect the output of the Link-in node to the input of the Split node, then the Split node output to the input of the Join node.

Up to this point, we've established a flow that creates a new property in customer data objects called 'full_name' and subsequently removes any unnecessary properties. In summary, we've cleaned the data according to our needs.

### Customer Data Deduplication

1. Drag a Change node onto the canvas, and set `msg.data` to `msg.payload` and `msg.payload` to `[]`.

> 2. Drag a MongoDB4 node onto the canvas, and configure it with your correct details. If you haven't used MongoDB with Node-RED, please refer to the [Using MongoDB with Node-RED](/blog/2024/04/using-mongodb-with-node-red/). Enter `find` into the Operation field.

Reviewer: I don't think MongoDB is used very often in an ETL context, precisely as it's structureless. Do we have a better example? An event based sensor perhaps?

{% endrenderFlow %}

1. To Deploy the flow, click on the top-right Deploy button
2. To visit the dashboard, Go to the Dashboard 2.0 sidebar and click on the top-right "Open dashboard" button.

!["Screenshot of dashboard visualizing process customer data based on job profile"](./images/etl-with-node-red-chart-based-on-job-profile.png "Screenshot of dashboard visualizing process customer data based on job profile"){data-zoomable}

!["Screenshot of dashboard visualizing process customer data based on age group"](./images/etl-with-node-red-chart-based-on-age.png "Screenshot of dashboard visualizing process customer data based on age group"){data-zoomable}

## What are the limitations of using Node-RED as an ETL tool

While Node-RED is versatile, there are some limitations to consider while using it as an ETL tool:

- Advanced Features: Some advanced ETL features, like automated schema detection and sophisticated error handling, might require additional customization or external modules.
- Data Governance: Node-RED does not inherently provide robust data governance and lineage tracking, which are often essential in enterprise ETL tools.

> - Scalability: While Node-RED can handle many heavy tasks, but it may not offer the same level of optimization for processing extremely large datasets compared to dedicated ETL tools.

Reviewer: ChatGPT?

Node-RED Dashboard 2.0 provides several chart options, and more are expected in the future. For more information and to explore available chart types, refer to the [Node-RED Dashboard 2.0 charts documentation](https://dashboard.flowfuse.com/nodes/widgets/ui-chart.html).

## Benefits of using Node-RED as an ETL tool.

1. **Ease of Use:** Its visual programming interface makes it accessible to non-developers.
2. **Flexibility:** A wide range of nodes and the ability to write custom JavaScript allow for flexible data processing.
3. **Integration:** Node-RED excels in integrating IoT devices and handling real-time data, making it well-suited for combining diverse data sources into unified workflows.
4. **Cost-Effective:** Being open-source, Node-RED can be a cost-effective alternative to expensive ETL tools.
5. **Community Support:** A large community provides a wealth of nodes, examples, and support.

## Building a Simple ETL Project Using Node-RED

> Let's walk through a simple project where we will use Node-RED as an ETL tool to extract sample customer data from an API, transform it, and then load cleaned data into a MongoDB database and process data into a local file.

Reviewer: I don't think we should advice to use MongoDB? PG, Snowflake, and others are more often a target for the data.

> *Note: The goal of this project is to understand how to utilize Node-RED as an ETL tool. We assume that the reader has a basic knowledge of Node-RED.*

Reviewer: ```suggestion
```

## Prerequisite 

Before proceeding further make sure you have installed the following nodes:

1. [node-red-contrib-mongodb4](https://flows.nodered.org/node/node-red-contrib-mongodb4)
2. [@flowfuse/node-red-dashboard](https://dashboard.flowfuse.com/)

## Extracting Data

1. Drag an Inject node onto the canvas.

> 2. Drag an HTTP Request node onto the canvas, double-click on it, and set the URL to `https://api.slingacademy.com/v1/sample-data/files/customers.json`.

Reviewer: This feels like cheating. Not a real world scenario
